app/controllers/cases/offender_sar_controller.rb

```diff
@@ -187,5 +191,28 @@ def preserve_step_state
     def session_state
       "#{@correspondence_type_key}_state".to_sym
     end
+
+    def load_optional_flags_from_params
+      @creation_optional_flags.each do |key, _|
+        @creation_optional_flags[key] = params[key]
+      end
+    end
+
+    def build_url_params_from_flags
+      if has_optional_flags?
+        "?#{@creation_optional_flags.to_param}"
+      else
+        ""
+      end
+    end
+
+    def has_optional_flags?
+      @creation_optional_flags.present? && @creation_optional_flags.values.all? {|x| x.present?}
+    end
+  
+    def back_link_url
+      "#{@case.case_route_path}/#{@case.get_previous_step}#{build_url_params_from_flags}"
+    end
+
```

Comment:
Wondering if some of this ought to be in a decorator? Perhaps OffenderBaseDecorator?

Reply:
@Polsonby,  actually I just moved it out of the decorator class 😳 as in order to construct the proper link I need to a few variables e.g. case_route_path and flags which are only available on a Controller level.   In a sense,  I prefer to have this link URL defined at controller level rather than decorator level, for me decorator is mainly used to enrich the ActiveRecord Model,  but anything to do with URL and behavior are in the controller